src/Fixer/Phpdoc/AlignMultilineCommentFixer.php

```diff
@@ -0,0 +1,165 @@
+use PhpCsFixer\FixerConfiguration\FixerConfigurationResolver;
+use PhpCsFixer\FixerConfiguration\FixerOptionBuilder;
+use PhpCsFixer\FixerDefinition\CodeSample;
+use PhpCsFixer\FixerDefinition\FixerDefinition;
+use PhpCsFixer\Tokenizer\Token;
+use PhpCsFixer\Tokenizer\Tokens;
+
+/**
+ * @author Filippo Tessarotto <zoeslam@gmail.com>
+ */
+final class AlignMultilineCommentFixer extends AbstractFixer implements ConfigurationDefinitionFixerInterface, WhitespacesAwareFixerInterface
+{
+    private $tokenKinds;
+
+    /**
+     * {@inheritdoc}
+     */
+    public function configure(array $configuration = null)
+    {
+        parent::configure($configuration);
+
+        $this->tokenKinds = [T_DOC_COMMENT];
+        if ($this->configuration['multiline_comment_blocks']) {
+            $this->tokenKinds[] = T_COMMENT;
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getDefinition()
+    {
+        return new FixerDefinition(
+            'Multiline doc comment: enforce asterisk start on each line [PSR-5] and align them.',
+            [
+                new CodeSample(
+'<?php
+    /**
+            * This is a DOC Comment
+with a line not prefixed with asterisk
+
+   */'
+                ),
+                new CodeSample(
+'<?php
+    /*
+            * This is a doc-like multiline comment
+*/',
+                    ['multiline_comment_blocks' => true]
+                ),
+                new CodeSample(
+'<?php
+    /*
+            * This is a doc-like multiline comment
+with a line not prefixed with asterisk
+
+   */',
+                    [
+                        'multiline_comment_blocks' => true,
+                        'multiline_comment_blocks_mixed_content' => true,
+                    ]
+                ),
+            ]
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isCandidate(Tokens $tokens)
+    {
+        return $tokens->isAnyTokenKindsFound([T_COMMENT, T_DOC_COMMENT]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function applyFix(\SplFileInfo $file, Tokens $tokens)
+    {
+        $lineEnding = $this->whitespacesConfig->getLineEnding();
+        $lineEndingLength = mb_strlen($lineEnding);
+        $indentation = false;
+        foreach ($tokens as $index => $token) {
+            $content = $token->getContent();
+            $newlinePosition = mb_strrpos($content, $lineEnding);
+
+            if (false === $indentation && false !== $newlinePosition) {
+                $indentation = mb_substr($content, $newlinePosition + $lineEndingLength);
+                continue;
+            }
+
+            if (false !== $indentation && isset($content[0]) && trim($content) === '') {
+                $indentation = preg_replace('/[ \t]*[\r\n]+/m', '', $content);
+                continue;
+            }
+
+            if (false !== $indentation && $newlinePosition && $token->isGivenKind($this->tokenKinds)) {
```

Comment:
Why would you need a backward loop? Checking if the previous token is a whitespace with a newline should be enough.

On a semantical point of view, your code says _"check all tokens to detect indentation, and if a comment happens to be found, apply this indentation to it"_. I would refactor the code so it says _"when finding a comment token, detect indentation and apply it to the token"_ instead.

I tried to rewrite `applyFix()`:
```php
    protected function applyFix(\SplFileInfo $file, Tokens $tokens)
    {
        $lineEnding = $this->whitespacesConfig->getLineEnding();
        foreach ($tokens as $index => $token) {
            if (!$token->isGivenKind($this->tokenKinds)) {
                continue;
            }

            $whitespace = '';
            $previousIndex = $index - 1;
            if ($tokens[$previousIndex]->isWhitespace()) {
                $whitespace = $tokens[$previousIndex]->getContent();
                --$previousIndex;
            }
            if ($tokens[$previousIndex]->isGivenKind(T_OPEN_TAG)) {
                $whitespace = preg_replace('/[^\s]/', '', $tokens[$previousIndex]->getContent()).$whitespace;
            }

            if (!preg_match('/\R([ \t]*)$/', $whitespace, $matches)) {
                continue;
            }

            $indentation = $matches[1];
            $lines = preg_split('/\R/', $token->getContent());
            $replaceCurrentCommentToken = true;
            foreach ($lines as $lineNumber => $line) {
                if (0 === $lineNumber) {
                    continue;
                }

                $line = ltrim($line);
                if ($token->isGivenKind(T_COMMENT) && (!isset($line[0]) || '*' !== $line[0])) {
                    if ('all_multiline' !== $this->configuration['comment_type']) {
                        $replaceCurrentCommentToken = false;
                        break;
                    }
                    continue;
                }

                if (!isset($line[0])) {
                    $line = '*'.$line;
                } elseif ('*' !== $line[0]) {
                    $line = '* '.$line;
                }

                $lines[$lineNumber] = $indentation.' '.$line;
            }

            if ($replaceCurrentCommentToken) {
                $tokens[$index] = new Token([$token->getId(), implode($lineEnding, $lines)]);
            }
        }
    }
```
As you can see, it only checks if the previous token is a whitespace (or an open tag, due to the way PHP tokenizes it). This also removes the need to use functions like `(mb_)strpos`. All tests pass. WDYT?